### PR TITLE
Chfast progpow

### PR DESCRIPTION
**Pull request by ed2k**
make progpow coexist with ethash to prepare for ethereum based code
base algorithm written by https://github.com/ifdefelse
functional example https://github.com/bitcoinnewyork/Bitcoinnyc, https://github.com/EtherCoinGo/ParityETE

### CODE REVIEW

File: lib/ethash/keccakf800.c [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+	0x0000800a, 0x8000000a, 0x80008081, 0x00008080, 0x80000001, 0x80008008
+};
+
+#define ROTL(x,n,w) (((x) << (n)) | ((x) >> ((w) - (n))))
+#define ROTL32(x,n) ROTL(x,n,32)	/* 32 bits word */
+
+void keccak_f800_round(uint32_t st[25], const int r)
+{
+
+	const uint32_t keccakf_rotc[24] = {
```
Comment (chfast) [inline lib/ethash/keccakf800.c]: These value are obviously wrong, probably for Keccak-f[1600].

Comment (ed2k) [reply lib/ethash/keccakf800.c]: which ones? keccakf_rndc[] is for 32bit keccak_800 not 1600

Comment (chfast) [reply lib/ethash/keccakf800.c]: Exactly. You can rol 32-bit integer by 36 bits. These values are probably taken from Keccak-f[1600] while they should be from Keccak-f[800] spec.

**Comment by chfast** [comment]
Hey @ed2k,

I added a benchmark for verify_progpow() to this PR. It indicates some problem in the implementation because I get:

```
----------------------------------------------------------------
Benchmark                         Time           CPU Iterations
----------------------------------------------------------------
verify                      1121222 ns    1120848 ns        564
verify_progpow                 1356 ns       1356 ns     513596
```

**Comment by ed2k** [comment]
yes, we know it should be very slow, still no idea that kind of slow. the
GPU version isn't that kind of gap.

**Comment by chfast** [comment]
Would you like to rebase this some day?

**Comment by ed2k** [comment]
rebase could be time consuming, I will try to upload a new one with some of
recent changes merged with up to date master.

File: lib/ethash/ethash.cpp [hunk 2]
```diff
@@ -137,50 +246,74 @@ void build_light_cache(hash512 cache[], int num_items, const hash256& seed) noex
...
+    hash1024 n1,n2;
+    n1 = calculate_dataset_item(context, index*2);
+    n2 = calculate_dataset_item(context, index*2+1);
+    return hash2048{{n1.hashes[0], n1.hashes[1], n2.hashes[0], n2.hashes[1]}};
+}
+
+/// Calculates a full l1 dataset item
+///
+/// This consist of one 32-bit items produced by calculate_dataset_item_partial().
+uint32_t calculate_L1dataset_item(const epoch_context& context, uint32_t index) noexcept
```
Comment (chfast) [inline lib/ethash/ethash.cpp]: Where is this specified?

**Comment by ed2k** [comment]
this is progpow's dag, which is just 2x of the original ethash's dag
https://github.com/ifdefelse/ProgPOW/blob/master/libprogpow/ProgPow.cpp#L78

**Pull request closed without merging**